### DISCUSSION TITLE
metro-serializer-esbuild causing slower app TTI (TTFD)

### DISCUSSION THREAD

**Discussion by BrentonWheeler** [microsoft/rnx-kit#2766] [Q&A]
I'm finding that when using `metro-serializer-esbuild` instead of just `metro-serializer` that, while my app JS bundle is reduced by 40% (awesome BTW), my app launch time has gone from ~5s to ~9s (on a slow device) and a similar almost doubling of launch time on faster devices (only tested Android so far). Tested via the "[Fully drawn](https://developer.android.com/topic/performance/vitals/launch-time#retrieve-TTFD)" `logcat` log value (which can be added via use of https://www.npmjs.com/package/react-native-startup-time)

~Let me know if you see this too otherwise I can look into making a reproduction example.~

EDIT: Reproduction example: https://github.com/BrentonWheeler/esbuild-tti-test-app

`yarn bundle:install` (using`metro-serializer`):

![Screen Shot 2023-10-24 at 11 42 13 am](https://github.com/microsoft/rnx-kit/assets/11394076/64007de4-d5f5-4963-be1c-9f8623ffdc93)

`yarn bundle:install+esbuild` (using `metro-serializer-esbuild`):

![Screen Shot 2023-10-24 at 11 40 16 am](https://github.com/microsoft/rnx-kit/assets/11394076/ca139c48-4f58-44d0-97d0-ea9c11afdf32)

**Comment by tido64** [comment]
Hi @BrentonWheeler, I think the difference you're seeing is due to `inlineRequires: true` in `metro.config.js`. If you disable it, what times do you see?

If I understand it correctly, `inlineRequires` basically moves your `require()` statements inline so that:

```js
const fs = require("fs");

function read(p) {
  return fs.readFile("package.json", "utf-8");
}
```

Becomes:

```js
function read(p) {
  return require("fs").readFile("package.json", "utf-8");
}
```

This can have a positive impact on your TTI because parts of your bundle is not executed until it is needed. However, we've disabled this in our apps because it assumes none of your dependencies have any side effects. We've experienced crashes because of this flag, so we always disable it now. This is also why we haven't seen this perf difference.

That said, the reason `inlineRequires` is disabled when you use `metro-serializer-esbuild` is because tree shaking only works when your code is in ESM. `inlineRequires` currently does not work on ESM, so it wouldn't have had any effect even if it was enabled. If you know which parts of your code is causing the slow down, you can manually inline `import()` those parts. Otherwise, I'm not aware of any good workarounds for this.

**Reply by BrentonWheeler** [reply]
Hi @tido64, thanks for that information although I'm getting the same (better) performance with `inlineRequires: false` when running `yarn bundle:install` (non-esbuild bundle) 🤔

**Reply by tido64** [reply]
That is interesting. Are you able to profile the two bundles to see where the time is spent? I am not able to reproduce it locally since I don't have a slow device lying around. I get around ~400 ms for either config.

**Reply by BrentonWheeler** [reply]
I'll look into that, but just checking - are you also unable to reproduce a difference in performance on a larger scale application with these 2 configs (and/or a slow Android simulator)?

**Reply by tido64** [reply]
I will have to check with someone else about that. No one internally have reported any slowdowns due to this, as far as I am aware. 

@crazyfraggle, did you check our TTI numbers when integrating this plugin?

**Reply by crazyfraggle** [reply]
In a way, yes. We had a small expected regression - about 200-400ms I think - since we lost support for RAM bundles. That is for the pure JS bundles though. For the ones were we use Hermes, I don't think we registered any regression.

**Reply by BrentonWheeler** [reply]
I'm experiencing this regression with hermes and without any other variables changing (like RAM bundles enabled in only one test, as shown in the reproduction repository). Are you able to reproduce with that repo or a larger app with those configs, and/or with a slow Android device/Simulator?

The performance regression is more clear the slower the device and the larger the bundle size

**Reply by BrentonWheeler** [reply]
I've had someone else reproduce this performance discrepancy with a slow Android simulator (Nexus 4) they saw 550ms average without esbuild vs 850ms average with esbuild. 300ms (35%) increase in TTI on this small isolated example repository, while reducing the bundle size (which should have a **_reduction_** on the TTI), really seems like something else is off here

**Reply by tido64** [reply]
It is really strange. I don't have a lot of devices lying around. The ones I do have are fairly new and fast. It would really help if you could profile the app on startup.

@crazyfraggle, I don't suppose you (or someone on your team) can reproduce this? It sounds like it should impact all of us.

**Reply by BrentonWheeler** [reply]
Let me know if this isn't exactly what you need, all I can tell from this is the `mqt_js` section in the middle takes a lot longer with esbuild 

profiles: `<removed>`

**Reply by BrentonWheeler** [reply]
Also somehow doing the following I am able to both have tree shaking and keep the performance of not using an esbuild bundle:
- Using the `esbuildTransformerConfig` but changing `experimentalImportSupport: true` from `false`
- Changing `disableImportExportTransform: false` from `true` in `babel.config.js`: 

```
presets: [
      ['@rnx-kit/babel-preset-metro-react-native', {
        disableImportExportTransform: false,
      }]
    ]
```

I don't think this is a real solution but I'm hoping it helps debug what is causing the performance degradation

**Reply by tido64** [reply]
This does help (or at least I think it does)! If you diff the bundles before and after the changes, you'll see there's a significant change in the output.

`experimentalImportSupport: true` the imports you have at the top i.e.:

```js
import * as md from 'react-icons/md'
import * as pi from 'react-icons/pi'
import * as gi from 'react-icons/gi'
...
export default function App() {
```

`App.js` is "hoisted" to the top level:

```js
  var md = __toESM(require_md());
  var pi = __toESM(require_pi());
  var gi = __toESM(require_gi());
  ...
  function App() {
    console.log(md);
    console.log(pi);
    console.log(gi);
```

Whereas with it off, `App.js` becomes a "CJS module":

```js
  var require_App = __commonJS({
    "App.js"(exports) {
      ...
      var md = _interopRequireWildcard(require_md());
      var pi = _interopRequireWildcard(require_pi());
      var gi = _interopRequireWildcard(require_gi());
      ...
```

If I understand this correctly, this means that at startup, all assets are imported before they are even used. This does not happen when `App.js` is CJS. It would explain the big slowdown that you see on slower devices.

If you inline the imports in `App.js`, you should see TTI improve:

```diff
diff --git a/App.js b/App.js
index b526701..65ec220 100644
--- a/App.js
+++ b/App.js
@@ -3,40 +3,40 @@ import React from 'react';
 import { StyleSheet, Text, View } from 'react-native';
 import { StartupTime } from 'react-native-startup-time';

-import * as md from 'react-icons/md'
-import * as pi from 'react-icons/pi'
-import * as gi from 'react-icons/gi'
-import * as ci from 'react-icons/ci'
-import * as fa from 'react-icons/fa'
-import * as fa6 from 'react-icons/fa6'
-import * as io from 'react-icons/io'
-import * as io5 from 'react-icons/io5'
-import * as ti from 'react-icons/ti'
-import * as go from 'react-icons/go'
-import * as fi from 'react-icons/fi'
-import * as lu from 'react-icons/lu'
-import * as wi from 'react-icons/wi'
-import * as di from 'react-icons/di'
-import * as ai from 'react-icons/ai'
-import * as bs from 'react-icons/bs'
-import * as ri from 'react-icons/ri'
-import * as fc from 'react-icons/fc'
-import * as gr from 'react-icons/gr'
-import * as hi from 'react-icons/hi'
-import * as hi2 from 'react-icons/hi2'
-import * as si from 'react-icons/si'
-import * as sl from 'react-icons/sl'
-import * as bi from 'react-icons/bi'
-import * as cg from 'react-icons/cg'
-import * as vsc from 'react-icons/vsc'
-import * as tb from 'react-icons/tb'
-import * as tfi from 'react-icons/tfi'
-import * as rx from 'react-icons/rx'
-import * as lia from 'react-icons/lia'
+export default async function App() {
+  // NOTE: These console.log's are just "using" the packages so they can't be trivally removed by metro without esbuild's treeshaking

+  const md = await import('react-icons/md');
+  const pi = await import('react-icons/pi');
+  const gi = await import('react-icons/gi');
+  const ci = await import('react-icons/ci');
+  const fa = await import('react-icons/fa');
+  const fa6 = await import('react-icons/fa6');
+  const io = await import('react-icons/io');
+  const io5 = await import('react-icons/io5');
+  const ti = await import('react-icons/ti');
+  const go = await import('react-icons/go');
+  const fi = await import('react-icons/fi');
+  const lu = await import('react-icons/lu');
+  const wi = await import('react-icons/wi');
+  const di = await import('react-icons/di');
+  const ai = await import('react-icons/ai');
+  const bs = await import('react-icons/bs');
+  const ri = await import('react-icons/ri');
+  const fc = await import('react-icons/fc');
+  const gr = await import('react-icons/gr');
+  const hi = await import('react-icons/hi');
+  const hi2 = await import('react-icons/hi2');
+  const si = await import('react-icons/si');
+  const sl = await import('react-icons/sl');
+  const bi = await import('react-icons/bi');
+  const cg = await import('react-icons/cg');
+  const vsc = await import('react-icons/vsc');
+  const tb = await import('react-icons/tb');
+  const tfi = await import('react-icons/tfi');
+  const rx = await import('react-icons/rx');
+  const lia = await import('react-icons/lia');

-export default function App() {
-  // NOTE: These console.log's are just "using" the packages so they can't be trivally removed by metro without esbuild's treeshaking
   console.log(md);
```

**Reply by BrentonWheeler** [reply]
> Whereas with it off, App.js becomes a "CJS module"

> If I understand this correctly, this means that at startup, all assets are imported before they are even used. This does not happen when App.js is CJS. It would explain the big slowdown that you see on slower devices.

Although I've found that `experimentalImportSupport: true` and `disableImportExportTransform: false` was better TTI than the opposite so I'm not sure if I'm understanding correctly 🤔 

But it's an interesting line of thought though, I don't suppose there is a way to transform all static `import`s (and/or `require`s) to be dynamic inlined `import`s via a build process so I could test this theory in a larger app to see if it holds up?

**Reply by tido64** [reply]
> Although I've found that `experimentalImportSupport: true` and `disableImportExportTransform: false` was better TTI than the opposite so I'm not sure if I'm understanding correctly 🤔

When `experimentalImportSupport: true` and `disableImportExportTransform: false`, Metro will perform its own transformations. Last I checked, Metro will transform `import`/`export` statements into `require()` statements. esbuild does not tree shake `require()` statements.

> But it's an interesting line of thought though, I don't suppose there is a way to transform all static `import`s (and/or `require`s) to be dynamic inlined `import`s via a build process so I could test this theory in a larger app to see if it holds up?

Not that I know of. `inlineRequires` was meant to do that, but it doesn't support ESM syntax.

**Reply by BrentonWheeler** [reply]
> When experimentalImportSupport: true and disableImportExportTransform: false, Metro will perform its own transformations. Last I checked, Metro will transform import/export statements into require() statements i.e. making them CJS modules.

Ah right makes sense, 1 thing I'm unclear on is why `experimentalImportSupport: true` and `disableImportExportTransform: false` isn't the default if it keeps tree shaking and avoids worsening TTI perf?

**Reply by BrentonWheeler** [reply]
Ah I missed this point which would clear up my last question:

> I checked, Metro will transform import/export statements into require() statements. **_esbuild does not tree shake require() statements._**

However I'm fairly sure I was having tree shaking work fine with this configuration 🤔 (`experimentalImportSupport: true` and `disableImportExportTransform: false`)

**Reply by tido64** [reply]
I'm not quite sure how Metro's import/export plugin works. If you enable it (`experimentalImportSupport: true` and `disableImportExportTransform: false`), it does produce slightly bigger bundles, but still smaller than without esbuild. Tbh, it's been a while since I've looked at this code, so I don't remember all the details but I do believe the difference was significant enough for our apps.

**Reply by BrentonWheeler** [reply]
FWIW I've largely resolved this issue via using [@babel/plugin-transform-modules-commonjs](https://babeljs.io/docs/babel-plugin-transform-modules-commonjs) and returning true from lazy for everything (ATM) eg:

```
plugins: [
        [
          '@babel/plugin-transform-modules-commonjs',
          {
            lazy: (source) => {
              return true;
            },
          },
        ],
...
```

It has allowed use of esbuild treeshaking for smaller bundle size, while maintaining the old TTI

**Reply by tido64** [reply]
Thanks for sharing!

I'm a little curious about how this works. It looks like `@babel/plugin-transform-modules-commonjs` is already a part of `@react-native/babel-preset`:

![image](https://github.com/microsoft/rnx-kit/assets/4123478/8c6fc10a-6a99-4a15-a602-6f7dd24f8129)
-- https://github.com/facebook/react-native/blob/0.73-stable/packages/react-native-babel-preset/src/configs/main.js#L65-L81

And it looks like it is included as long as `disableImportExportTransform` is falsy. We encourage users to enable this to disable transforming `import` -> `require` as this will prevent esbuild from tree shaking.

**Reply by BrentonWheeler** [reply]
Is it somehow possible that the treeshaking is done then the remaining smaller bundle is converted?

**Reply by tido64** [reply]
Unless something has changed recently, tree shaking is the very last step, just before writing to disk. That's why it's so important to preserve import/export statements till the very end. Both Babel and Metro configs are there to ensure this. We also had to make changes in Metro itself to make this possible.

**Reply by chang-ke** [reply]
> FWIW I've largely resolved this issue via using [@babel/plugin-transform-modules-commonjs](https://babeljs.io/docs/babel-plugin-transform-modules-commonjs) and returning true from lazy for everything (ATM) eg:
> 
> ```
> plugins: [
>         [
>           '@babel/plugin-transform-modules-commonjs',
>           {
>             lazy: (source) => {
>               return true;
>             },
>           },
>         ],
> ...
> ```
> 
> It has allowed use of esbuild treeshaking for smaller bundle size, while maintaining the old TTI

Hey @BrentonWheeler , would you mind to share the whole configuration? as I'm facing the same issue, thanks!

**Reply by BrentonWheeler** [reply]
Hey @chang-ke, it's just the reproduction repo but with that change to the `babel.config.js` shown above, eg: https://github.com/BrentonWheeler/esbuild-tti-test-app/pull/1/files

**Reply by chang-ke** [reply]
@BrentonWheeler this plugin indeed make the TTI faster, but the bundle size also increased..  causing tree shaking failed

**Reply by BrentonWheeler** [reply]
Hmm, that is inline with what @tido64 has said so you both may be right, but on the reproduction repo I'm getting the same bundle size with and without the plugin:

```
yarn bundle:analyse = Bundle is 46.55 MB in size
yarn bundle:analyse+esbuild (without the plugin) = Bundle is 34.7 MB in size 
yarn bundle:analyse+esbuild (with the plugin) = Bundle is 34.7 MB in size 
```
